### ISSUE TITLE
VS Code Files

### ISSUE THREAD

**Issue by beverts312** [swellaby/generator-swell#9]
I was thinking about adding support to optionally add VS Code files (to the express generator only to start with) so that user could have preconfigured shortcuts for building, debugging etc.

Any thoughts on if we should either:  
* Add Optionally  
* Always add  
* Never add  

?

**Comment by calebcartwright** [comment]
Love it! I would say make it optional for any Atom/Sublime lovers out there. Someone was telling me the other day that you can configure it to install extensions for the user too

**Comment by beverts312** [comment]
sounds good
possibly in the future we can change it to a choice input and support additional IDE's like those

**Comment by calebcartwright** [comment]
Definitely. Would that be good as a standalone generator (we can always leverage composeWith in this and other generators) or will some of the items in .vscode for example vary a lot between baed on user input?

**Comment by beverts312** [comment]
The tasks.json should be the same for all of them or atleast it would share a common base.
The launch.json could vary significantly (i.e. vsts build task wouldnt have one, express would look similar to cli but they would look vastly different from angular/aurelia).  

Originally I was thinking to add it in its own sub folder under `templates`, but for it to have it to be handled in code for the other generators. May have to tinker with it a bit to get a clean structure without overengineering it.  

Thoughts/suggestions?

**Comment by calebcartwright** [comment]
think there are a few ways we can tackle this. the question I have been using to determine where a template file is placed within the directory structure is whether it will be more difficult long term to have to modify a template file replicated in multiple directories, or to maintain large numbers of variables within a file along with the code needed to insert the values at scaffold time. 

do you have any examples of what those two files look like?

if tasks.json is nearly identical then I think we can have a single place where we maintain the template file (and add a couple variables as needed). i'd recommend multiple buckets for launch.json based on the anticipated significant of variance.

I originally created the conditionals directory for content that could ultimately be "promptable" but would be (essentially) the same across all project types. But it may make more sense to have a .vscode subdir under each of the template directories

**Comment by beverts312** [comment]
Here is an example for a web app:  
```json  
{
    "version": "0.1.0",
    "configurations": [
        {
            "name": "Launch localhost with sourcemaps",
            "type": "chrome",
            "request": "launch",
            "url": "http://localhost/mypage.html",
            "webRoot": "${workspaceRoot}/app/files",
            "sourceMaps": true
        },
        {
            "name": "Launch index.html (without sourcemaps)",
            "type": "chrome",
            "request": "launch",
            "file": "${workspaceRoot}/index.html"
        },
    ]
}
```
Here is an example for a node app:  
```json  
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "node",
            "request": "launch",
            "name": "Launch Program",
            "program": "${workspaceRoot}/app.js",
            "cwd": "${workspaceRoot}"
        },
        {
            "type": "node",
            "request": "attach",
            "name": "Attach to Process",
            "port": 5858
        }
    ]
}
```

**Comment by beverts312** [comment]
The proper implementation of this feature being blocked by an issue I am seeing in glob with respect to ignoring files in `.dot` folders (I commented with details here:  [node-glob 309](https://github.com/isaacs/node-glob/issues/309) )

**Comment by calebcartwright** [comment]
I should probably remember this, but does the glob parameter itself accept an array or is it just a string? If the former then perhaps we could just add a !pattern to the array to exclude/ignore?

**Comment by beverts312** [comment]
just a string unfortunatley

**Comment by calebcartwright** [comment]
@swellaby/core-collaborators - Any votes on whether or not this should be a candidate for the 1.0?

**Comment by beverts312** [comment]
ya it probably should be, debugging is an integral piece of the dev loop
I will work on this over the weekend

**Comment by calebcartwright** [comment]
@beverts312 - currently the [scaffolded .gitignore file is configured to exclude the .vscode ](https://github.com/swellaby/generator-swell/blob/master/generators/app/templates/boilerplate/gitignore#L9) directory and contents. Should that be removed/updated based on the content you are adding as part of this issue?

**Comment by calebcartwright** [comment]
Also, here is info around the extensions I referenced previously https://code.visualstudio.com/docs/editor/extension-gallery#_workspace-recommended-extensions Should that be included with this, or perhaps track that as a separate issue?

**Comment by beverts312** [comment]
I dont think we should force any extension on people, the only time I think it would be appropriate is if we added support for some project type that had a special debugger (i.e angular2 we would want the chrome debugger) then we could forceably install that

and ya personally I would not exclude it from source control, but I dont know if I am in the majority on that one

**Comment by calebcartwright** [comment]
I don't think it is possible to force/auto-install extensions on behalf of the user. However, what it does do is populate the list of extensions that are displayed when the user (well after the generator completes and exits) selects the 'Show recommended extensions' from within VS Code.

I think the extensions.json file is something that will be useful for folks collaborating as it is a really easy/accurate way to recommend extensions to each other. I don't know if we should necessarily have the generator populate the file, apart from perhaps something simple & direct like your example, but my thought was that we could have the generator include the file (since we are already creating the .vscode dir with other content). Could also add some info in the scaffolded readme and/or the docs of our generator itself.

@beverts312  does that change anything or do you still think we should leave it out?

**Comment by beverts312** [comment]
@calebcartwright  ahh got it, I think we should include it but track it separately so we can have a conversation about what we want to put in it

**Comment by calebcartwright** [comment]
on the gitignore for vscode, I started excluding it a long time ago back when it was essentially only used for local user/workspace configuration as I thought it would be common for users to have varying preferences that should be excluded. 

but now that appears to have changed quite dramatically so I am thinking that the template gitignore should **NOT** exclude the .vscode directory. Users can always update it down the line (post scaffolding) if their circumstances change. 

I will update the gitignore file since I will already be in that area addressing #48

**Comment by calebcartwright** [comment]
and agreed on extensions.json, will create a new issue
